### DISCUSSION TITLE
How to use proxy_protocol with nginx-proxy?

### DISCUSSION THREAD

**Discussion by bluepuma77** [nginx-proxy/nginx-proxy#1621] [Q&A] [answered]
Is is possible to enable [nginx proxy_protocol](https://docs.nginx.com/nginx/admin-guide/load-balancer/using-proxy-protocol/) with nginx-proxy? If yes, how (config files, env variables, etc)?

There is a TCP load balancer in front of my nginx-proxy and I would like to know the original IP of the request. 

I didn't find any information in the readme, previous issues or via web search.

**Comment by bluepuma77** [comment]
No one using proxy protocol with nginx-proxy and can tell how this works?

**Comment by talha13** [comment]
There are no default support for proxy protocol in nignx-proxy. But you can do this by updating nginx.tmpl file and then mount this to nginx-proxy container. Following things need to be changed on nginx.tmpl file

1. Change log format

``
log_format vhost '$host $remote_addr - $remote_user [$time_local] '
                 '"$request" $status $body_bytes_sent '
                 '"$http_referer" "$http_user_agent"';
``

to this

``
log_format vhost '$host $proxy_protocol_addr - $remote_user [$time_local] '
                 '"$request" $status $body_bytes_sent '
                 '"$http_referer" "$http_user_agent"';
``

2.  Change proxy header

``
proxy_set_header X-Real-IP $remote_addr;
``
``
proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
``

to this

``
proxy_set_header X-Real-IP       $proxy_protocol_addr;
``
``
proxy_set_header X-Forwarded-For $proxy_protocol_addr;
``

3. Set real_ip_header, set_real_ip_from directive

``
real_ip_header proxy_protocol;
``

``
set_real_ip_from xxx.xxx.xxx.xxx; # it should be your load balancer ip
``

4. Add the proxy_protocol parameter to the listen directive. There are around 12 listen directive in nginx.tmpl file

``
listen {{ $external_http_port }} {{ $default_server }};
``

to this

``
listen {{ $external_http_port }} {{ $default_server }} proxy_protocol;
``

5. Mount your updated nginx.tmpl file to the container 

``
/your/nginx.tmpl:/app/nginx.tmpl
``


You can check sample nginx.tmpl file from this link [Sample nginx.tmpl](https://github.com/talha13/nginx.tmpl.git)

Also you can check the nginx official documentation [Documentation](https://docs.nginx.com/nginx/admin-guide/load-balancer/using-proxy-protocol/)

**Comment by buchdag** [comment, marked as answer]
Support for Proxy Protocol has been added in #2587 

The doc is here : https://github.com/nginx-proxy/nginx-proxy/tree/main/docs#proxy-protocol-support
